csa/man/acctdusg.8.in
.\"
.\" (C) 2004-2008 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This documentation supports a
.\" program that is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" The documentation that supports
.\" this program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this documentation and the software it
.\" supports, are distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\"
.\" Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
.\" Sunnyvale, CA  94085, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan
.\"
.TH acctdusg 8
.SH NAME
\f4acctdusg\f1 \- reads standard input and computes disk resource consumption
.SH SYNOPSIS
.nf
\f4@sbindir@/acctdusg \f1[\f4-u\f1 \f2file\f1] [\f4-p\f1 \f2file\f1]\f1
.fi
.SH DESCRIPTION
\f4acctdusg\fP takes input from standard input and produces a report of disk
consumption sorted by user ID. The input to this command normally comes from
output of "find" command. For example:
.PP
.nf 
.RS .5i
.ft 4
# find /var |acctdusg    
00000   root             126184
00002   daemon                8
00003   adm                 264
00004   lp                    8
00009   news                 24
00025   named                24
00029   rpcuser             136
00037   rpm              101672
00038   ntp                   8
00051   smmsp                 8
00067   webalize             16
00080   desktop             112
#
.ft 1
.RE
.fi
.PP 
The output from this command  consists of three column of data: User login ID, login name, 
and number of disk blocks consumed by the user including indirect blocks.
.PP
If \f4-u\f1 option is specified, the disk consumption information on files  
that acctdusg failed to charge to any valid login are saved in the          
specified \f3unchargable_file\f1. This option is useful to find out which users
are avoiding disk charges.                                    
.PP                                                           
If \f4-p\f1 option is specified, the provided password file is used by      
acctdusg in finding login information. When \f4-p\f1 option is not specified,
the system call \f4getpwent\f1 is used to retrieve user account information.
.SH FILES
.nf
.ta 19
\f4/etc/passwd\fP       used for login name to user ID conversions
\f4@sbindir@/\fP   holds all administration accounting commands
\f4@csalocalstatedir@/day/pacct\fP        current process accounting file
.fi
.SH REFERENCES
chkconfig(8),
csaaddc(8),
dodisk(1M).


csa/ChangeLog
* Wed Jun  8 2016 Jay Lan <jay.j.lan@nasa.gov>
- Converted csa package from the last SGI-supported version 4.1.1 to
  community version 5.0.0

* Fri Oct 17 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.1
- Various fixes to allow for SLES11 build

* Fri Jul 25 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.0
- Fixed memory leaks
- Changed acctjob structure offset type from int to int64_t

* Fri Apr 18 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.2
- Updated to use new libnl API
- Fixed process exit code assignment

* Tue Mar 4 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.1
- Cleaned up acctdusg
- Added check to prevent SIGPIPE when communicating with daemon
- Increased maximum PID allowed for larger systems

* Wed Nov 7 2007 Jonathan Lim <jlim@sgi.com>
- csa-4.0.0
- Implement kernel functionality as a userspace daemon
- Remove obsolete wait metrics and replace with taskstats delay accounting
- Various bug fixes

* Fri Apr 27 2007 Jay Lan <jlan@sgi.com>
* Mon Apr 30 2007 Jonathan Lim <jlim@sgi.com>
- csa-3.0.2
- SBU's data was all zero in csa-3.0.0 and csa-3.0.1
- Clarify csacom(1) man page on '-f' option
- Obsolete csacom '-a' & '-q' option
- Fix a problem of csadrep reading input from a binary file
- Fixed csadrep(8) man page: replaced '-w' with '-n' and removed '-t'

* Wed Jul 5 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.1
- 'csacom -b' was broken in csa-3.0.0
- 'csacom -b' drops the last record of a pacct file

* Thu Feb 24 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.0
- csacom -[esES] option was broken.
- create /usr/sbin/rccsa symlink at %install in the spec file
- remove code for CSA record types not supported in Linux
- spliting up csa_kern.h into kernel version and userland version
- support pacct files created in csa-2.*.*

* Wed Jul 27 2005 Chris Sturtivant <csturtiv@sgi.com>
- csa-2.2.2
- Disabled perror for permission denied in csa_ctl when called by csa_auth

* Mon May 2 2005 Jay Lan <jlan@sgi.com>
- csa-2.2.1
- clean up csa init script
- csa_wracct(3) man page update
- 'csacom -o' wrote only one record
- 'csacom -b' does not work
- csajrep man page references invalid option '-W'
- 'csajrep -r' does not work
- 'dodisk' created files with incorrect user and group name 'adm'
- CSA does not handle gid's > 65535
- 'csacom -j' failed if job id greater than max value of an int

* Mon Nov 22 2004 Jay Lan <jlan@sgi.com>
- csa-2.2.0
- Fixed compilation issue with gcc 3.3.3
- CSA assumes existence of 'adm' user/group account
- csacom & csacms usage otuput missing carriage return
- csarecy misses a carriage return on certain condition
- csacom -W segfaults after printing one bad time

* Fri Oct 8 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.2
- Fixed 'csaswitch -c halt' bug
- Fixed 'csaedit dumped core on bad wkmg record' bug

* Tue Jul 27 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.1
- Remove static API library libcsa.a from rpm package
- Fixed a bug at csacom with invalid -u option
- cascom headers and data not aligned correctly at output

* Wed Jun 30 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.0
- Fixed a problem dealing with accounting files > 2GB
- Now derives "Day of Year" from other fields of the holidays file
- Removed "Pages Swapped" from csacom and csacms - not supported in Linux
- Fixed a bug of not mapping UID 0 to root

* Tue Feb 17 2004 Jay Lan <jlan@sgi.com>
- csa-2.0.0
- Implemented /proc ioctl interface
- Renamed rpm from csa-job-acct to csa
- Renamed /etc/init.d/csaacct to /etc/init.d/csa

* Fri Nov 7 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-2.0.0
- Provided API library libcsa.so and associated header file

* Wed Sep 3 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-1.0.3
- Updated the spec file

* Thu Feb 13 2003 Robin Holt <holt@sgi.com>
- Updated init scripts to use /var/lock/subsys

* Tue Nov 5 2002 Richard Offer <offer@sgi.com>
- Fixed packager/distribution/vendor

* Thu May 23 2002 Robin Holt <holt@sgi.com>
- csa-job-acct-1.0.2
- Changed install destination from /usr/local to /usr
- Updated the man page location to /usr/share/man
- Added a clean and realclean target to the Makefile

* Thu Apr 5 2001 Marlys Kohnke <kohnke@sgi.com>
- Changed csadrep internal markers back to match those in IRIX CSA
- Changed ja summary report so that all lines are displayed even if the
  resource used values are zero

* Sun Feb 11 2001 Marlys Kohnke <kohnke@sgi.com>
- Added support for Workload Management daemon accounting records
- Fixed csachkpacct script problem (df -k needs to be df -kP; changed awk
  field to 4)

* Fri Jan 19 2001 Daryl Grunau <dwg@acl.lanl.gov>
- port to Linux from SGI IRIX
- csa-job-acct-1.0.0 (available January 19, 2001)
- initial csa-job-acct-1.0.0 open source release


csa/cmd/acctdusg.c
/*
 * CSA acctdusg - Compute disk usage by login
 *
 * Copyright (c) 2000-2002 Silicon Graphics, Inc and LANL  All Rights Reserved.
 * Copyright (c) 2004-2008 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
 * Sunnyvale, CA  94085, or:
 *
 * http://www.sgi.com
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif  /* HAVE_CONFIG_H */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include "acctdef.h"
#include "acctmsg.h"
 
/* Globals */
#define FILENAME_SIZE 2048
#define USER_NAME_LENGTH 32  /* login.c */

/* struct to hold in disk usage info for a given user */
struct du {
	char uname[USER_NAME_LENGTH];
	uid_t uid;
	uint64_t blocks;
	struct du *next;
};

struct du *du_list = NULL; /* head of the link list of du's */


/* prototypes */
static void openError(char * f);
static void chargeFileOwner(char *, FILE *);
static int fileExists(char * fname);
static void printDiskUsage(void);

/* stub prototypes */
static int create_du_list(struct passwd *p);
static void free_du_list(void);
static void dsort ();
static int strndx(char * s, char c);
static	void	swapd(struct du *, struct du *, struct du *);


/* main */

int main (int argc, char **argv) {
	char pwFileName[FILENAME_SIZE+1] = "";  /* Name of /etc/passwd */
	char unchargedListName[FILENAME_SIZE+1] = "";  /* name of file w/list of
							  unaccounted files */
	char unchargedBak[FILENAME_SIZE+5] = ""; /* for rename() to .bak */
	int c; /* used for getopt */
	FILE	*pwFile=NULL;  /* file pointer for altenrnate pw file */
	FILE *unchargedListFile=NULL;
	struct passwd *pwEntry;
	char lineBuf [BUFSIZ+1];
	int retVal, err = 0;

	/* Command line arguement processing */
	/* getopt(3) man page from ProPack 2.4 used as a reference here */
	while (1) {
		c = getopt (argc, argv, "u:p:");
		if (c == -1)
			break;

		switch (c) {
		case 'u':
			strncpy(unchargedListName, optarg, FILENAME_SIZE);
			if (strlen(unchargedListName) == 0 ||
			    unchargedListName[0] == '-') {
			    	fprintf(stderr, "Command Usage: /usr/sbin/acctdusg [-u file] [-p file]\n");
				exit(1);
			}
			/* printf("DEBUG uncharged file list output file: %s\n", unchargedListName); */
			break;
		case 'p':
			strncpy(pwFileName, optarg, FILENAME_SIZE);
			/* printf("DEBUG alternate pw file specified: %s\n", pwFileName); */
			break;
		default:
			fprintf(stderr, "Command Usage: /usr/sbin/acctdusg [-u file] [-p file]\n");
			exit(1);
		}
	}
	/* We always want files 0644.... */
	umask(022);

	/* end of command line arguement processing */

	if (strlen(unchargedListName) != 0) {
		/* An uncharged file list was specified */
		/* Are we already open?  This shouldn't happen. */
		if (unchargedListFile != NULL) {
			fprintf(stderr, "ERROR - file already open: %s\n", unchargedListName);
			exit(1);
		}
		if (fileExists(unchargedListName) == 0) {
			/* Since the file exists, we need to back it up to .bak */
			strncpy(unchargedBak, unchargedListName, FILENAME_SIZE+5);
			strcat(unchargedBak, ".bak"); /* We know we have room for .bak */
			retVal = rename(unchargedListName, unchargedBak);
			if (retVal != 0) {
				fprintf(stderr, "Error renaming %s to %s.  Quitting.\n",
					unchargedListName, unchargedBak);
				exit(1);
			}
		}

		/* Finally, we can open the darn thing for write*/
		unchargedListFile = fopen(unchargedListName, "w");
		if (unchargedListFile == NULL) {
			perror("Unable to open file for -u option...");
			openError(unchargedListName);
		}
		/* We will be writing to this file later. */

	}

	if (strlen(pwFileName) != 0) {
		/* an alternative password file was supplied with -p */
		if (pwFile != NULL) {
			/* File is already open - reset it */
			rewind(pwFile);
		}
		else {
			/* open the file */
			pwFile = fopen(pwFileName, "r");
			if (pwFile == NULL) {
				openError(pwFileName);
			}
		}
		/* Now we are ready to process the user-supplied password file */
		while (!err && (pwEntry = fgetpwent(pwFile)) != NULL) {
			err = create_du_list(pwEntry);
		}
		fclose(pwFile);
		pwFile=NULL;
	}
	else {
		/* No alternative password file specified, use standard pw lookups */
		setpwent();

		/* Process available password entries using getpwent */
		while (!err && (pwEntry = getpwent()) != NULL) {
			err = create_du_list(pwEntry);
		}
		endpwent();
	}

	if (!err)
		dsort();

	/* Start reading from stdin and charge users for file usage */

	while (!err && (fgets(lineBuf, BUFSIZ, stdin)) != NULL) {
		retVal = strndx(lineBuf, '\n');
		/* The string should terminate at the new line  - erasing the line feed */
		if (retVal != -1) {
			lineBuf[retVal] = '\0';
		}
		chargeFileOwner(lineBuf, unchargedListFile);
	}

	if (unchargedListFile != NULL) {
		fclose(unchargedListFile);
	}

	if (!err)
		printDiskUsage();

	free_du_list();

	return(err);
}

/* fileExists:
 * Check if a file exists.  Return 0 if it exists.  Return 1 if it doesn't.
 * Return -1 if we got some error.
 */
static int fileExists(char * fname) {
	struct stat finfo;
	int retVal;
	int MyErrno;
	
	retVal = stat(fname, &finfo);
	MyErrno = errno;
	
	if (retVal == 0) {
		return (0); /* file exists, return 0 */
	}
	else if (MyErrno == ENOENT) {
		return(1); /* file doesn't exist */
	}
	else {
		perror("stat returned an error not handled by fileExists");
		return(-1);
	}
}

/* printDiskUsage:
 *
 * This function prints out disk usage information from the linked list of
 * users.  It simply prints the output to stdout.
 */
static void printDiskUsage(void) {
	struct du *du_entry;

	if (du_list == NULL) {
		fprintf(stderr, "Error - no users exist?  This shouldn't happen.\n");
		exit(1);
	}

	du_entry = du_list;

	while (du_entry != NULL) {
		if (du_entry->blocks != 0) {
			printf("%05u\t%-8.8s\t%7lu\n", du_entry->uid, du_entry->uname, 
				du_entry->blocks);
		}
		du_entry = du_entry->next;
	}
}

/* openError:
 * This function reports an error opening a file and exits.
 * As intput, it takes the filename (string) that we attempted to open.
 */
static void openError(char *f) {
	fprintf(stderr, "ERROR - cannot open file: %s\nExiting.\n", f);
	exit (1);
}

/* chargeFileOwner:
 *
 * This function updates the linked list of users with file usage 
 * information.
 *
 * If there are files for which we can't find owners, we write usage 
 * information to a file specified with the -u option to this program.
 * If the -u option wasn't given, we produce no such output file.
 *
 */
static void chargeFileOwner(char * fname, FILE * fptr) {
	struct stat finfo;  /* used to store struct from stat */
	struct du *du_entry;
	int retVal;

	retVal = lstat(fname, &finfo);
	if (retVal != 0) {
		return;
	}

	/* Now go through the linked list looking for the user */
	if (du_list == NULL) {
		fprintf(stderr, "Error - no users exist?  This shouldn't happen.\n");
		exit(1);
	}

	du_entry = du_list;

	while (du_entry != NULL) {
		if(finfo.st_uid == du_entry->uid) {
			/* Found the user in the list, charge that user! */
			du_entry->blocks += finfo.st_blocks;
			return;
		}
		du_entry = du_entry->next;
	}
	/* If we made it here, it means we couldn't match the owner of the file */

	if (fptr != NULL) {
		/* Ok.  We need to store orphan usage data in this file */
		fprintf(fptr, "%5u\t%7lu\t%s\n", finfo.st_uid, finfo.st_blocks, fname);
		return;
	}
}

/* create_du_list:
 *
 * This function populates the disk usage struct list (du_list) with
 * info from getpwent (account info).  It's basically used to create
 * the struct for each user on the system.  Later, this list will be
 * used to store associated disk space usage information.
 */
static int create_du_list(struct passwd *p) {
	struct du *du_new, **du_entry;

	/* printf("DEBUG - create_du_list got pw entry for: %s\n", p->pw_name); */

	/* Allocate and populate new entry. */

	du_new = (struct du *) malloc(sizeof(struct du));	
	if (du_new == NULL) {
		perror("Failed to allocate memory for disk usage structure du");
		return -1;
	}
	strcpy(du_new->uname, p->pw_name); /* bounds checked by main */
	du_new->uid = p->pw_uid;
	du_new->blocks = 0;
	du_new->next = NULL;

	/* Append new entry to the list. */

	for (du_entry = &du_list; *du_entry != NULL; du_entry = &(*du_entry)->next)
		;
	*du_entry = du_new;

	return 0;
}

static void free_du_list(void) {
	struct du *curr, *next;

	for (curr = du_list; curr != NULL; curr = next) {
		next = curr->next;
		free(curr);
	}

	return;
}


static int
strndx(char *str, char chr)
{

	int index;

	for (index=0; *str; str++,index++)
		if (*str == chr)
			return index;
	return -1;
}

/*
 *	sort by increasing uid
 */
static void
dsort()
{

	struct du *dp1, *dp2, *pdp;
	int	change;

	if(du_list == NULL || du_list->next == NULL)
		return;

	change = 0;
	pdp = NULL;

	for(dp1 = du_list; ;) {
		if((dp2 = dp1->next) == NULL) {
			if(!change)
				break;
			dp1 = du_list;
			pdp = NULL;
			change = 0;
			continue;
		}
		if(dp1->uid > dp2->uid) {
			swapd(pdp, dp1, dp2);
			change = 1;
			dp1 = dp2;
			continue;
		}
		pdp = dp1;
		dp1 = dp2;
	}
}

static void
swapd(struct du *p, struct du *d1, struct du *d2)
{
	struct du *t;

	if (p) {
		p->next = d2;
		t = d2->next;
		d2->next = d1;
		d1->next = t;
	} else {
		t = d2->next;
		d2->next = d1;
		d1->next = t;
		du_list = d2;
	}
}

